src/runtime_src/core/pcie/noop/shim.cpp

```diff
@@ -108,6 +104,173 @@ free(unsigned int handle)
 
 } // buffer
 
+// Model the programming part of the device.
+// Pretend a device that supports multiple xclbins, which can can be
+// loaded by host application into different slots, where slots are
+// chosen by the driver (this code).
+namespace pl {
+
+class device
+{
+  // model multiple partitions
+  using slot_id = xrt_core::device::slot_id;
+  std::map<slot_id, xrt::xclbin> xclbins;
+
+  // capture cu data based on which slot it is associated with
+  struct cu_data {
+    std::string name;  // cu name
+    slot_id slot = 0;  // slot in which this cu is opened
+    uint32_t ctx = 0;  // how many contexts are opened on the cu
+  };
+  std::map<uint32_t, cu_data> idx2cu;  // idx -> cu_data
+
+  // cu indices are per device from 0..127.
+  // keep a stack of indices that can be used
+  // once last context on cu is released, its index is recycled
+  static constexpr uint32_t cu_max = 128;
+  std::vector<uint32_t> free_cu_indices;  // push, back, pop
+
+  // slot index for xclbin is a running incremented index
+  slot_id slot_index = 0; // running index
+
+  // exclusive locking to prevent race
+  std::mutex mutex;
+
+  // assign a slot index to an xclbin
+  void
+  register_xclbin(slot_id slot, const xrt::xclbin& xclbin)
+  {
+    xclbins[slot] = xclbin;
+  }
+
+public:
+  // device ctor, initialize free cu indices
+  device()
+  {
+    free_cu_indices.reserve(128);
+    for (int i = 127; i >=0; --i)
+      free_cu_indices.push_back(i);
+  }
+
+  // assign next slot index to an xclbin
+  void
+  register_xclbin(const xrt::xclbin& xclbin)
+  {
+    std::lock_guard lk(mutex);
+    register_xclbin(slot_index++, xclbin);
+  }
+
+  // unload an xclbin (todo)
+  void
+  unload_xclbin(slot_id slot)
+  {
+    std::lock_guard lk(mutex);
+    xclbins.erase(slot);
+  }
+
+  // open context on cu
+  // return: 0 on success, EAGAIN if xclbin was re-loaded, else error
+  int
+  open_context(slot_id slot, const xrt::uuid& xid, const std::string& cuname, bool shared)
+  {
+    std::lock_guard lk(mutex);
+
+    // this xclbin must have been registered
+    const auto& xclbin = xclbins[slot];
+    if (!xclbin)
+      throw xrt_core::error("Slot xclbin mismatch, no such registered xclbin in slot: " + std::to_string(slot));
+
+    // find cu in xclbin
+    auto cu = xclbin.get_ip(cuname);
+    if (!cu)
+      throw xrt_core::error("No such cu: " + cuname);
+
+    // Current user-level implementation only attempts opening of context
+    // once per CU.  In other words, it is an error if this function is
+    // called twice on same CU within same process and since this noop driver
+    // is tied to a process we simply throw.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-variable"
+    for (const auto& [idx, cud] : idx2cu)
+      if (cud.name == cuname && cud.slot == slot)
+        throw xrt_core::error("Context already opened on cu: " + cuname);
+#pragma GCC diagnostic pop
+#if 0
```

Comment:
Is this code intentionally left behind.  If so, a commend describing why and when it should be used would be helpful.